feat: assertions rules (#489)
Redocly/openapi-cli: docs/resources/built-in-rules.md

OpenAPI-cli supports [custom rules](./custom-rules.md).
However, if you have an idea for a built-in rule you believe will benefit the greater API community, please [open an issue](https://github.com/Redocly/openapi-cli/issues/new).

## Generic built-in rules

You can use special `enforcements` rule for cases which our other built-in rules do not cover.

> There are [a set of generic built-in rules](#list-of-generic-built-in-rules) which can be configured under `enforcements` to enforce different generic usecases.

Reviewer: Won't it be confusing to have rules inside a rule? How about calling them `enforcements`? @skadinna @adamaltman

Reviewer: I think assertions is best

### How it works?

Let's say you want to validate Operation summary and Info description fields - they should be defined, not empty, 

> minimum 30 chars length and both have "_hello_" substring inside.

Reviewer: I would use some more realistic example here. `_hello_` substring is not a realistic use case. How about "should end with a full stop"?

Reviewer: Then, nonEmpty and minimum 30 chars are redundant. You need only 30 chars.

> For these usecases you can create next enforcements:

Reviewer: Isn't it a single usecase?

```yaml
lint:
  rules:
    enforcements:
      - on
        - Operation.summary
        - Info.description

>         description: Summary and Description fields don't match enforcements

Reviewer: ```suggestion
        description: Operation summary and description fields should be at least 30 characters and end with a full stop.
```

        severity: error
        defined: true
        nonEmpty: true
        minLength: 30
        regex: /hello/
```

In other words, it generates 8 unique custom rules `operation-summary-defined`, `operation-summary-non-empty`, `operation-summary-min-length`, `operation-summary-regex`,

> `info-description-defined`, `info-description-non-empty`, `info-description-min-length`, `info-description-regex`.

Reviewer: This part is confusing. It doesn't generate 8 unique custom rules technically. Also what are those names? We don't have such rules. I would remove it.

> ### Format of `enforcements` rule

Reviewer: ```suggestion
### Format of enforcements
```

Basically `enforcements` rule is an array of enforcements, where each one has the next fields structure:

> - `on` - the path or array of the paths to the properties on which we should do the linting. The path is the string splitted by `.`.

Reviewer: This is not easy to understand unfortunately . We need to rewrite it.

It's not clear what should be put into on. We need to link to the list of node types.

@skadinna @adamaltman could you help?

> - `description` - the message which shows in case of fail. Not required.

Reviewer: ```suggestion
- `description` - optional error message.
```

- `severity` - the severity level of particular case (`off`, `warn`, `error`). Not required, by default `error`.
- the [list of generic rules](#list-of-generic-built-in-rules) with their options.

> ## List of generic built-in rules

Reviewer: Again, shouldn't be called rules.

Reviewer 2: If you don't mind, I'll make a commit adjusting these docs.

### enum

To enforce a value to be among set of predefined values.

```yaml
lint:
  rules:
    enforcements:

>       - on Operation.summary

Reviewer: ```suggestion
      - on: Operation.summary
```

        description: Summary should be amoung predefined values
        severity: error
        enum:
          - one
          - two
```

### pattern

To enforce a value to match a regex.

```yaml
lint:
  rules:
    enforcements:

>       - on Operation.summary

Reviewer: ```suggestion
      - on: Operation.summary
```

        description: Summary should match a regex
        severity: error
        pattern: /test/
```

### casing

To enforce specific casing style (camelCase, kebab-case, snake_case, PascalCase)

```yaml
lint:
  rules:
    enforcements:

>       - on Operation.operationId

Reviewer: ```suggestion
      - on: Operation.operationId
```

        description: Operation Id should be camelCase
        severity: error
        casing: camelCase
```

### mutuallyExclusive 

To mark some properties as mutually exclusive (`a` or `b` but not `a` and `b` together).
This rule runs only on node. It means that the `on` path should end with the node, but not the property.
In case if the last part of the path is the property, the rule will be ignored.

```yaml
lint:
  rules:
    enforcements:

>       - on Operation

Reviewer: ```suggestion
      - on: Operation
```

        description: Operation should not have summary and security fileds together
        severity: error
        mutuallyExclusive:
          - summary
          - security
```

### mutuallyRequired

To mark some properties as mutually required (`a` and `b` but not only `a` or only `b`).
This rule runs only on node. It means that the `on` path should end with the node, but not the property.
In case if the last part of the path is the property, the rule will be ignored.

```yaml
lint:
  rules:
    enforcements:

>       - on Operation

Reviewer: ```suggestion
      - on: Operation
```

        description: Operation should have summary and security fileds together
        severity: error
        mutuallyRequired:
          - summary
          - security
```

### defined

To check if some property is defined. 

```yaml
lint:
  rules:
    enforcements:

>       - on Operation.operationId

Reviewer: ```suggestion
      - on: Operation.operationId
```

        description: Operation Id should be defined
        severity: error
        defined: true
```

### undefined

To check if some property is undefined.

```yaml
lint:
  rules:
    enforcements:

>       - on Operation.summary

Reviewer: ```suggestion
      - on: Operation.summary
```

        description: Operation summary should be undefined
        severity: error
        undefined: true
```

### nonEmpty

To check if some property is not empty.

```yaml
lint:
  rules:
    enforcements:

>       - on Operation.summary

Reviewer: ```suggestion
      - on: Operation.summary
```

        description: Operation summary should not be empty
        severity: error
        nonEmpty: true
```

### length

To enforce length of specific array or string.

```yaml
lint:
  rules:
    enforcements:

>       - on Operation.summary

Reviewer: ```suggestion
      - on: Operation.summary
```

        description: Operation summary should have 20 chars length
        severity: error
        length: 20
```

### minLength

To enforce minimal length of specific array or string.

```yaml
lint:
  rules:
    enforcements:

>       - on Operation.summary

Reviewer: ```suggestion
      - on: Operation.summary
```

        description: Operation summary should have minimum 20 chars length
        severity: error
        minLength: 20
```

### maxLength

To enforce maximal length of specific array or string.

```yaml
lint:
  rules:
    enforcements:

>       - on Operation.summary

Reviewer: ```suggestion
      - on: Operation.summary
```

        description: Operation summary should have maximum 20 chars length
        severity: error
        maxLength: 20
```

### sortOrder

To enforce sort order in array of strings or array of objects (collection). 
If the `property` fields is present, it means that we check the sort order direction on collection. 
Also, in case of array of strings the direction could be also set like `sortOrder: asc`.

```yaml
lint:
  rules:
    enforcements:

>       - on Operation.tags

Reviewer: ```suggestion
      - on: Operation.tags
```
